ext/cache/01_cache.js

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+"use strict";
+
+((window) => {
+  const core = window.__bootstrap.core;
+  const {
+    PromiseResolve,
+    PromiseReject,
+    TypeError,
+    Uint8Array,
+  } = window.__bootstrap.primordials;
+  const {
+    toInnerResponse,
+    toInnerRequest,
+  } = window.__bootstrap.fetch;
+  const { getHeader } = window.__bootstrap.headers;
+
+  class CacheStorage {
+    constructor() {
+      return this;
+    }
+
+    async open(cacheName) {
+      const cacheId = await core.opAsync("op_cache_storage_open", cacheName);
+      return PromiseResolve(new Cache(cacheId));
+    }
+
+    async has(cacheName) {
+      return await core.opAsync("op_cache_storage_has", cacheName);
+    }
+
+    async delete(cacheName) {
+      return await core.opAsync("op_cache_storage_delete", cacheName);
+    }
+  }
+
+  class Cache {
+    #id;
+
+    constructor(cacheId) {
+      this.#id = cacheId;
+      return this;
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#dom-cache-put */
+    async put(request, response) {
+      // Step 1.
+      let innerRequest = null;
+      // Step 2.
+      if (request instanceof Request) {
```

Comment:
iirc, the usual idiom is
```suggestion
      if (ObjectPrototypeIsPrototypeOf(RequestPrototype, request) {
```